### PR TITLE
Ruby on Rails capstone project - Budget app

### PR DESCRIPTION
**Pull request by medaminedev66**
## presentation:
watch this [video](https://youtu.be/psSkAXMT4K0)
## For this project I have done the following:
- Used ruby gems as software packages system.
- Installed Ruby on Rails framework.
- Implemented Rails RESTful design and router.
- Used controllers to handle requests and render empty views.
- Used params from browser request in a safe way.
- Used preprocessed html file with embedded Ruby code.
- Used layouts and templates for shared content.
- Used database migration files to maintain database schema.
- Used validations for models.
- Secured app from n+1 problems.
- Wrote SQL queries with ActiveRecord.
- Set up associations between models.
- Built a webapp that requires the user to log in.
- Used devise gem for authentication.
- Limited access to webapp resources based on authorization rules.

### CODE REVIEW

File: app/assets/stylesheets/users.scss [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+@import "variables";
+@import "globals";
+
+.signup {
+  margin-top: 5rem;
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  //   border: 1px solid gray;
+  background-color: rgb(241, 241, 241);
```
Comment (Emmyn5600) [inline app/assets/stylesheets/users.scss]: - [ ] Remove the commented code.

File: app/assets/stylesheets/users.scss [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+.home {
+  background-color: rgb(241, 241, 241);
+  height: 100%;
+  width: 100%;
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+  //   align-items: center;
+}
+
```
Comment (Emmyn5600) [inline app/assets/stylesheets/users.scss]: - [ ] Remove the commented code

File: app/controllers/activities_controller.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
+class ActivitiesController < ApplicationController
+  def index
+    @category = Group.find_by(id: params[:group_id])
+    @transactions = Activity.includes(:groups).where("groups.id = #{params[:group_id]}").references(:groups)
+    @total_amount = 0
+    @transactions.each do |activity|
+      @total_amount += activity.amount
+    end
+  end
```
Comment (Emmyn5600) [inline app/controllers/activities_controller.rb]: - [ ] We don't implement SQL Queries inside the controller, we use the model to keep the SQL queries implementation.

File: app/controllers/activities_controller.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+  def new
+    @activity = Activity.new
+    @groups_list = Group.where(user_id: current_user.id)
+  end
```
Comment (Emmyn5600) [inline app/controllers/activities_controller.rb]: - [ ] We don't implement SQL Queries inside the controller, we use the model to keep the SQL queries implementation.

File: app/controllers/activities_controller.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    @activity = Activity.new(activity_params)
+    @activity.author_id = current_user.id
+    if @activity.save
+      flash[:success] = 'Object successfully created'
+      redirect_to activities_home_path(@activity.groups.first.id)
+    else
+      flash[:error] = 'Something went wrong'
+      render 'new'
+    end
+  end
```
Comment (Emmyn5600) [inline app/controllers/activities_controller.rb]: - [ ] We don't implement SQL Queries inside the controller, we use the model to keep the SQL queries implementation.

**Review by Emmyn5600 — CHANGES_REQUESTED** [review]
Hi @medaminedev66 

Great work on making the project requirements 👏 
You've done well implementing some of the project requirements, but there are still some which aren't addressed yet, please work on it and ask for another review.

Highlights:

- The project is running successfully
- You have a descriptive Readme file
- Your database is well structured
- You have implemented almost all requirements.
- You have descriptive PR
- All tests are passing successfully.
      
 ## Required Changes ♻️

- _Check the comment under the review._

## Optional suggestions

- n/a

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### _**Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.**_
------------------------------------------------------------------------------------------------------------------------------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using this [form.](https://airtable.com/shrQAqnBwek5a0O0s)

File: app/models/activity.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+class Activity < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  has_and_belongs_to_many :groups
+
+  validates :name, presence: true, length: { maximum: 50 }
+  validates :amount, numericality: { greater_than_or_equal_to: 0 }
+end
```
Comment (enaburekhan) [inline app/models/activity.rb:7]: **Optional**
Nice job here. However according to the requirement in the `transaction page`: For a given category, the list of transactions is presented ordered by the most recent. Currently, it's the reverse. See screen-shot below.

image

I will suggest you fix this before merging the PR.

**Review by enaburekhan — APPROVED** [review]
Hi Medaminedev66, 

Thank you for making the requested changes. Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Code review by **Eric Enaburekhan**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/activity.rb (hunk 6 above)
Comment (medaminedev66) [reply app/models/activity.rb:7]: Thanks for reminding me about that. I will do it. 👍

**Pull request merged**